src/fund/shares/Shares.sol

```diff
@@ -251,68 +251,54 @@ contract Shares is IShares, TokenUser, Spoke, SharesToken {
-    }
-
-    /// @notice Redeem a specified quantity of the sender's shares
-    /// for ONLY SPECIFIC ASSETS in the fund, forfeiting the remaining assets
-    /// @dev Do not call directly, unless an asset throws preventing redemption.
-    /// Calling directly with a limited set of assets will result in the sender
-    /// losing claim to the remaining assets for their shares. It is intended as
-    /// a last resort for users to directly redeem their assets.
-    /// @dev Rewards all fees prior to redemption
-    /// @param _sharesQuantity The amount of shares to redeem
-    /// @param _assets The assets to receive from the redemption
-    function redeemSharesWithConstraints(uint256 _sharesQuantity, address[] memory _assets)
-        public
-    {
-        require(_assets.length > 0, "redeemSharesWithConstraints: _assets cannot be empty");
-        require(_sharesQuantity > 0, "redeemSharesWithConstraints: _sharesQuantity must be > 0");
-
-        IFeeManager(routes.feeManager).rewardAllFees();
-
         require(
             _sharesQuantity <= balanceOf(msg.sender),
-            "redeemSharesWithConstraints: _sharesQuantity exceeds sender balance"
+            "redeemSharesQuantity: _sharesQuantity exceeds sender balance"
         );
 
+        IVault vault = IVault(routes.vault);
+        address[] memory ownedAssets = vault.getOwnedAssets();
+        require(ownedAssets.length > 0, "redeemSharesQuantity: ownedAssets is empty");
+
+        IFeeManager(routes.feeManager).rewardAllFees();
+
         // Destroy the shares
         uint256 sharesSupply = totalSupply();
         _burn(msg.sender, _sharesQuantity);
 
-        // Calculate and transfer owed assets to redeemer
-        IVault vault = IVault(routes.vault);
-        uint256[] memory owedQuantities = new uint256[](_assets.length);
-        for (uint256 i = 0; i < _assets.length; i++) {
-            uint256 quantityHeld = vault.assetBalances(_assets[i]);
-            require(quantityHeld > 0, "Requested asset holdings is 0");
+        // Calculate and transfer owned assets to redeemer
+        uint256[] memory ownedQuantities = new uint256[](ownedAssets.length);
+        for (uint256 i = 0; i < ownedAssets.length; i++) {
+            uint256 quantityHeld = vault.assetBalances(ownedAssets[i]);
+            // Redeemer's ownership percentage of asset holdings
+            ownedQuantities[i] = mul(quantityHeld, _sharesQuantity) / sharesSupply;
 
-            // Confirm asset has not already been transfered
-            for (uint256 j = 0; j < i; j++) {
+            // Transfer owed asset to redeemer
+            try vault.withdraw(ownedAssets[i], ownedQuantities[i]) {}
+            catch {}
+
+            uint256 receiverPreBalance = IERC20(ownedAssets[i]).balanceOf(msg.sender);
+            try IERC20Flexible(ownedAssets[i]).transfer(msg.sender, ownedQuantities[i]) {
+                uint256 receiverPostBalance = IERC20(ownedAssets[i]).balanceOf(msg.sender);
```

Comment:
I don't think we need to validate the pre/post-transfer balances here, because there is nothing we can do about a fund asset that does not actually transfer tokens like it is supposed to. We only need to protect against an asset that is unredeemable.